src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -799,16 +799,87 @@ private void CopyToObjects(object[] array, int index)
         /// made to the dictionary after <see cref="GetEnumerator"/> was called.
         /// </remarks>
-        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
+        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator() => new Enumerator(this);
+
+        /// <summary>Provides an enumerator implementation for the dictionary.</summary>
+        private sealed class Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>
         {
-            Node?[] buckets = _tables._buckets;
-            for (int i = 0; i < buckets.Length; i++)
+            // Provides a manually-implemented version of (approximately) this iterator:
+            //     Node?[] buckets = _tables._buckets;
+            //     for (int i = 0; i < buckets.Length; i++)
+            //         for (Node? current = Volatile.Read(ref buckets[i]); current != null; current = current._next)
+            //             yield return new KeyValuePair<TKey, TValue>(current._key, current._value);
+
+            public readonly ConcurrentDictionary<TKey, TValue> _dictionary;
+
+            private ConcurrentDictionary<TKey, TValue>.Node?[]? _buckets;
+            private Node? _node;
+            private int _i;
+            private int _state;
+
+            private const int StateUninitialized = 0;
+            private const int StateOuterloop = 1;
+            private const int StateInnerLoop = 2;
+            private const int StateDone = 3;
+
+            public Enumerator(ConcurrentDictionary<TKey, TValue> dictionary)
             {
-                // The Volatile.Read ensures that we have a copy of the reference to buckets[i]:
-                // this protects us from reading fields ('_key', '_value' and '_next') of different instances.
-                for (Node? current = Volatile.Read(ref buckets[i]); current != null; current = current._next)
+                _dictionary = dictionary;
+                _i = -1;
+            }
+
+            public KeyValuePair<TKey, TValue> Current { get; private set; }
+
+            object IEnumerator.Current => Current;
+
+            public void Reset()
+            {
+                _buckets = null;
+                _node = null;
+                Current = default;
+                _i = -1;
+                _state = StateUninitialized;
+            }
+
+            public void Dispose() { }
+
+            public bool MoveNext()
+            {
+                switch (_state)
                 {
-                    yield return new KeyValuePair<TKey, TValue>(current._key, current._value);
+                    case StateUninitialized:
+                        _buckets = _dictionary._tables._buckets;
+                        _i = -1;
+                        goto case StateOuterloop;
+
+                    case StateOuterloop:
+                        ConcurrentDictionary<TKey, TValue>.Node?[]? buckets = _buckets;
+                        Debug.Assert(buckets != null);
```

Comment:
Stylistically, you could just `!` the nullability away instead as the assert adds nothing even in debug? (It will throw on the next line)?

Reply:
It could, but I wanted to highlight with an assert that it will definitively be non-null here, rather than just ignoring warnings about it.